packages/node/src/lib/log.ts

```diff
@@ -0,0 +1,145 @@
+
+  res.write = (chunk, encoding, cb) => {
+    res._body += chunk;
+    write.call(res, chunk, encoding, cb);
+  };
+
+  res.end = (chunk, encoding, cb) => {
+    // Chunk is optional in res.end
+    // http://nodejs.org/dist/latest/docs/api/http.html#http_response_end_data_encoding_callback
+    if (chunk) res._body += chunk;
+    end.call(res, chunk, encoding, cb);
+  };
+}
+
+export interface Options extends LogOptions {
+  bufferLength?: number;
+  baseLogUrl?: string;
+}
+
+/**
+ * This middleware will set up Express to automatically log all API requests to ReadMe Metrics.
+ *
+ * @param apiKey The API key for your ReadMe project. This ensures your requests end up in your dashboard. You can read more about the API key in [our docs](https://docs.readme.com/reference/authentication).
+ * @param group A function that helps translate incoming request data to our metrics grouping data. You can read more under [Grouping Function](#grouping-function).
+ * @param options Additional options. See the documentation for more details.
+ * @returns Your Express middleware
+ */
+export function log(
+  readmeApiKey: string,
+  req: Request,
+  res: ExtendedResponse,
+  group: GroupingObject,
+  options: Options = {}
+) {
+  if (!readmeApiKey) throw new Error('You must provide your ReadMe API key');
+  if (!group) throw new Error('You must provide a group');
+
+  // Ensures the buffer length is between 1 and 30
+  const bufferLength = clamp(options.bufferLength || config.bufferLength, 1, 30);
+
+  const baseLogUrl = options.baseLogUrl || undefined;
+
+  const startedDateTime = new Date();
+  const logId = uuidv4();
+
+  patchResponse(res);
+
+  if (baseLogUrl !== undefined && typeof baseLogUrl === 'string') {
+    res.setHeader('x-documentation-url', `${baseLogUrl}/logs/${logId}`);
+  }
+
+  function startSend() {
+    // This should in future become more sophisticated,
+    // with flush timeouts and more error checking but
+    // this is fine for now
+    const groupData = group;
+
+    const payload = constructPayload(
+      req,
+      // {
+      //   ...req,
+
+      //   // Shallow copying `req` destroys `req.headers` on Node 16 so we're re-adding it.
+      //   headers: req.headers,
+
+      //   // If you're using route nesting with `express.use()` then `req.url` is contextual to that route. So say
+      //   // you have an `/api` route that loads `/v1/upload`, `req.url` within the `/v1/upload` controller will be
+      //   // `/v1/upload`. Calling `req.originalUrl` ensures that we also capture the `/api` prefix.
+      //   url: req.originalUrl,
+      // },
+      res,
+      {
+        ...groupData,
+        logId,
+        startedDateTime,
+        responseEndDateTime: new Date(),
+        routePath: req.route
+          ? url.format({
+              protocol: req.protocol,
+              host: req.hostname,
+              pathname: `${req.baseUrl}${req.route.path}`,
+            })
+          : '',
+        responseBody: res._body,
```

Comment:
Definitely a little weird that it's `_body` instead of `body`, feels like maybe a future footgun. One option could be that when we build up the `_body` field we treat it like a temp variable and then just replace `body` with it and delete `_body`. Is it necessary for us to keep both around? If so, let's just maybe add some more comments for context.

Reply:
res._body and req.body are two different things. req.body is set by express's body-parser, and res._body is set by us as a place to store everything that's being sent down the wire as a response body. I opted to underscore prefix it to make it clearer that this is ours and not to be tampered with. Also this code has been like this for years, I didn't change this 😅

Reply:
Annoying this is coming up as a completely new file and not a rename. I renamed express-middleware.ts to log.ts. I probably made enough changes that git is just treating it as new though.